### PR TITLE
Event usage split clickhouse queries

### PR DESCRIPTION
**Pull request by timgl**
## Changes

- This makes sure all queries are executed in sequence rather than in parallel to avoid overwhelming clikhouse
- It executes all the event volume queries in one go
- It'll grab all property key usage of low volume teams in one go
- For large team usage we do a query for each, again sequentially

## Checklist

- [ ] All querysets/queries filter by Organization, Team, and User (if this PR affects ANY querysets/queries).
- [ ] Django backend tests (if this PR affects the backend).
- [ ] Cypress end-to-end tests (if this PR affects the frontend).

### CODE REVIEW

File: posthog/tasks/calculate_event_property_usage.py [hunk 1]
```diff
@@ -13,85 +14,140 @@
...
+    def _save_team(self, team: Team) -> None:
+        def _sort(to_sort: List) -> List:
+            return sorted(to_sort, key=lambda item: (item.get("usage_count", 0), item.get("volume", 0)), reverse=True)
+
+        team.event_names_with_usage = _sort([val for _, val in self._event_names_dict[team.pk].items()])
+        team.event_properties_with_usage = _sort([val for _, val in self._event_properties_dict[team.pk].items()])
+        team.save()
+
+    def calculate_usage_count_for_team(self, team: Team) -> None:
+        self._event_names_dict[team.pk] = {event: {"event": event, "usage_count": 0} for event in team.event_names}
```
Comment (macobo) [inline posthog/tasks/calculate_event_property_usage.py:54]: Semantics question: Do we care about items which have usage count of 0? Asking to propose a refactoring.

**Review by macobo — COMMENTED** [review]
The control flow here is quite hard to follow (with calculations and saving things intermixed) and a ton of code has try: catch thrown around to deal with the chosen data structure.

I think I have a proposal here but need a bit of context around semantics, see below

File: posthog/tasks/calculate_event_property_usage.py [hunk 2]
```diff
@@ -13,85 +14,140 @@
...
+            for event in item.filters.get("events", []):
+                if event["id"] in self._event_names_dict[team.pk]:
+                    self._event_names_dict[team.pk][event["id"]]["usage_count"] += 1  # type: ignore
+
+            for prop in item.filters.get("properties", []):
+                if prop.get("key") in self._event_properties_dict[team.pk]:
+                    self._event_properties_dict[team.pk][prop["key"]]["usage_count"] += 1  # type: ignore
+
+        # intermittent save in case the heavier queries don't finish
+        self._save_team(team)
```
[the remark this answers is not in the document]
Comment (timgl) [reply posthog/tasks/calculate_event_property_usage.py:68]: As per comment above the calculate event volume might fail, but at least we'll have the usage count set correctly.

File: posthog/tasks/calculate_event_property_usage.py (hunk 1 above)
Comment (timgl) [reply posthog/tasks/calculate_event_property_usage.py:52]: Yes we want those in this dict, as we use this dict to show the select box in the frontend

File: posthog/tasks/calculate_event_property_usage.py [hunk 3]
```diff
@@ -13,85 +14,138 @@
...
+
+        timestamp = now() - timedelta(days=30)
+
+        # Combine all low volume clients in one query
+        volume = sync_execute(
+            GET_PROPERTIES_LOW_VOLUME, {"timestamp": timestamp, "cutoff": self._clickhouse_volume_cutoff},
+        )
+        self.save_properties_volume(volume)
+
+        high_volume_clients = [
```
Comment (macobo) [inline posthog/tasks/calculate_event_property_usage.py:120]: Nit: This could be a set.

**Review by macobo — COMMENTED** [review]
So I tried refactoring this to make the control flow a bit cleaner by:
1. Extracting a team-based class that can be used to populate values/call save, with the main class responsible for sql queries + control flow
2. Using defaultdict under the hood to get rid of many of the try-catch clauses, constructing dicts over operating with a list of tuples.
3. Making `save` calls super explicit in the code.

All-in-all this came out the same length, but I think this is a bit cleaner. What do you think?

```python
from datetime import timedelta
from typing import Dict, List, Optional, Tuple, Union
from collections import defaultdict

from celery.app import shared_task
from django.db import connection
from django.db.models import Count
from django.db.models.query import QuerySet
from django.utils.timezone import now

from posthog.ee import is_ee_enabled
from posthog.models import Team
from posthog.models.dashboard_item import DashboardItem
from posthog.models.event import Event

ValueByTeamEvent = Dict[Tuple[number, str], number]

def calculate_event_property_usage() -> None:
    CalculateEventPropertyUsage().run()

def key_by_team_and_event(query_results: List[Tuple[int, str, int]]) -> ValueByTeamEvent:
    return { (team_id, event): value for team_id, event, value in query }

class CalculateEventPropertyUsage:
    _teams: QuerySet
    _event_names_dict: Dict[int, Dict[str, Dict[str, Union[int, str]]]] = {}
    _event_properties_dict: Dict[int, Dict[str, Dict[str, Union[int, str]]]] = {}

    def __init__(self, clickhouse_volume_cutoff: int = 1_000_000) -> None:
        # Number of events at which we query events individually
        # Only set in tests
        self._clickhouse_volume_cutoff = clickhouse_volume_cutoff
        self.team_calculators = [CalculateTeamEventPropertyUsage(team) for team in Team.objects.all()]

    def run(self) -> None:
        for team_calculator in self.team_calculators:
            team_calculator.populate_event_usage_count()
            team_calculator.save()

        event_volumes = self._get_events_volume()
        for team_calculator in self.team_calculators:
            team_calculator.populate_event_volume(event_volumes)
            team_calculator.save()

        if is_ee_enabled():
            self._clickhouse_calculate_properties_volume()
        else:
            volumes = self._psql_get_properties_volume()
            for team_calculator in self.team_calculators:
                team_calculator.populate_property_volume(volumes)
                team_calculator.save()

    def _psql_get_properties_volume(self) -> ValueByTeamEvent:
        timestamp = now() - timedelta(days=30)
        cursor = connection.cursor()
        cursor.execute(
            "SELECT team_id, json_build_array(jsonb_object_keys(properties)) ->> 0 as key1, count(1) FROM posthog_event WHERE timestamp > %s group by team_id, key1 order by count desc",
            [timestamp],
        )
        return key_by_team_and_event(cursor.fetchall())

    def _clickhouse_calculate_properties_volume(self) -> None:
        from ee.clickhouse.client import sync_execute
        from ee.clickhouse.sql.events import GET_PROPERTIES_FOR_TEAM, GET_PROPERTIES_LOW_VOLUME

        timestamp = now() - timedelta(days=30)

        # Combine all low volume clients in one query
        low_usage_volumes = key_by_team_and_event(sync_execute(
            GET_PROPERTIES_LOW_VOLUME, {"timestamp": timestamp, "cutoff": self._clickhouse_volume_cutoff},
        ))
        for team_calculator in self.team_calculators:
            team_calculator.populate_property_volume(volumes)
            team_calculator.save()

        high_volume_clients = set(
            team_id
            for team_id, _, _ in sync_execute(
                "SELECT team_id FROM (SELECT team_id, count(1) as count FROM events GROUP BY team_id) WHERE count >= {}".format(
                    self._clickhouse_volume_cutoff
                ),
                {"timestamp": timestamp},
            )
        )
        for team_calculator in self.team_calculators:
            if team_calculator.team.pk in high_volume_clients:
                volume = key_by_team_and_event(sync_execute(GET_PROPERTIES_FOR_TEAM, {"timestamp": timestamp, "team_id": team_calculator.team.pk},))
                team_calculator.populate_property_volume(volumes)
                team_calculator.save()

    def _get_events_volume(self) -> ValueByTeamEvent:
        timestamp = now() - timedelta(days=30)
        if is_ee_enabled():
            from ee.clickhouse.client import sync_execute
            from ee.clickhouse.sql.events import GET_EVENTS_VOLUME

            return key_by_team_and_event(sync_execute(GET_EVENTS_VOLUME, {"timestamp": timestamp},))
        return key_by_team_and_event(
            Event.objects.filter(timestamp__gt=timestamp)
            .values("team_id", "event")
            .annotate(count=Count("id"))
            .values_list("team_id", "event", "count")
        )
        team.save()

class CalculateTeamEventPropertyUsage:
    def __init__(self, team):
        self.team = team
        self.event_usage_count = defaultdict(int)
        self.event_properties_count = defaultdict(int)
        self.event_usage_volume = defaultdict(int)
        self.event_properties_volume = defaultdict(int)

    def populate_event_usage_count(self) -> None:
        for item in DashboardItem.objects.filter(team=self.team, created_at__gt=now() - timedelta(days=30)):
            for event in item.filters.get("events", []):
                self.event_usage_count[event["id"]] += 1

            for prop in item.filters.get("properties", []):
                if "key" in prop: # Note: Only needed if we need none protection
                    self.event_properties_count[prop["key"]] += 1

    def populate_event_volume(self, event_usage_volume: ValueByTeamEvent):
        self._populate(self.team.event_names, event_usage_volume, self.event_usage_volume)

    def populate_property_volume(self, property_usage_volume: ValueByTeamEvent):
        self._populate(self.team.event_properties, property_usage_volume, self.event_properties_volume)

    def _populate(self, keys: List[str], usage_dict: ValueByTeamEvent, counts_object: Dict[str, int]):
        for key in keys:
            counts_object[key] = usage_dict.get((team.pk, key), 0)

    def save(self):
        team.event_names_with_usage = self._construct_usage_volume_list(
            self.team.event_names,
            self.event_usage_count,
            self.event_usage_volume
        )
        team.event_properties_with_usage = self._construct_usage_volume_list(
            self.team.event_properties,
            self.event_properties_count,
            self.event_properties_volume
        )

    def _construct_usage_volume_list(self, keys, usage_counts, volumes):
        to_sort = [
            { "usage_count": usage_counts[key], "volume": volumes[key]}
            for key in keys
        ]

        return sorted(to_sort, key=lambda item: (item.get("usage_count", 0), item.get("volume", 0)), reverse=True)
```

**Review by macobo — APPROVED** [review]
:heart:
